### PR TITLE
improve pre-build target function checking

### LINKED ISSUE #27: Fix false negative 'function used in fuzz targets'
Our current code automatically fails an LLM-generated fuzz target if it [does not contain the function under test](https://github.com/google/oss-fuzz-gen/blob/main/experiment/builder_runner.py#L100C17-L100C42).
However, the current pattern-matching is native and has false negatives (i.e., function used correctly but not recognized by us), and rejects valid fuzz targets.

This relates to our function name parsing regex:
https://github.com/google/oss-fuzz-gen/blob/4e3a5ab4cc11df7dd4d99d57612c3c62c034975b/experiment/benchmark.py#L40-L43

A quick temporary solution is checking the function names without special chars.
Later we can make better use of data from FI

### PR DESCRIPTION
**Pull request by cjx10**
Fixes #27

### CODE REVIEW

**Comment by DonggeLiu** [comment]
Thanks!
Please post the experiment job and report link here once you have them :)

E.g., https://github.com/google/oss-fuzz-gen/pull/93#issuecomment-1937999076

File: experiment/builder_runner.py [hunk 1]
```diff
@@ -88,9 +88,12 @@ def _contains_target_function(self, target_path: str) -> bool:
...
     with open(target_path) as generated_code_file:
       generated_code = generated_code_file.read()
-    # Get the top level function name without namespace.
-    target_function_name = self.benchmark.function_name.rsplit('::', 1)[-1]
-    return bool(target_function_name in generated_code)
+    # Get the top level function name without namespace or params.
+    pattern = r'(?:[a-zA-Z_]\w*::)*([a-zA-Z_]\w*)(?:\s*<.*>)?\s*\('
+    match = re.search(pattern, self.benchmark.function_signature)
+    min_func_name = (match.group(1).strip()
+                     if match else self.benchmark.function_signature)
```
Comment (DonggeLiu) [inline experiment/builder_runner.py:95]: nit: Could you please split this into a new function?
The function takes the function signature and returns the minmized_function_name.

File: experiment/builder_runner.py [hunk 2]
```diff
@@ -88,9 +88,12 @@ def _contains_target_function(self, target_path: str) -> bool:
...
+    return bool(min_func_name in generated_code)
```
Comment (DonggeLiu) [inline experiment/builder_runner.py:96]: No need for `bool()`. `min_func_name in generated_code` already returns bool.

Comment (DonggeLiu) [reply experiment/builder_runner.py:96]: Also, line 110/113 below should use `function_signature` not `function_name`.

**Review by DonggeLiu — CHANGES_REQUESTED** [review]

File: experiment/builder_runner.py (hunk 1 above)
Comment (cjx10) [reply experiment/builder_runner.py]: done

File: experiment/builder_runner.py (hunk 2 above)
Comment (cjx10) [reply experiment/builder_runner.py]: done

File: experiment/builder_runner.py [hunk 3]
```diff
@@ -84,16 +84,20 @@ def _libfuzzer_args(self) -> list[str]:
         '-len_control=0'
     ]
 
+  def _get_minimum_func_name(self, func_sig: str) -> str:
+    """Extract the minimum function name from function signature,
```
Comment (DonggeLiu) [inline experiment/builder_runner.py:88]: nit: Extracts

Comment (cjx10) [reply experiment/builder_runner.py]: done

**Comment by cjx10** [comment]
JOB: https://pantheon.corp.google.com/kubernetes/job/us-central1/llm-experiment/default/ofg-pr-95-jim
REPORT: https://llm-exp.oss-fuzz.com/Result-reports/ofg-pr/2024-02-12-95-jim-comparison/index.html

**Review by DonggeLiu — APPROVED** [review]

**Pull request merged**